QDWH

adam-lewis-gcorp opened the pull request.

adam-lewis-gcorp committed:
QDWH

jakevdp commented:
Is this meant as a replacement for #6751?

adam-lewis-gcorp commented:
@googlebot I signed it!

adam-lewis-gcorp committed:
Merge branch 'google:main' into master

adam-lewis-gcorp committed:
Adds spectral divide and conquer eigensolver

Significant improvements to QDWH.

Mirror Scipy function signature.
Support both 'left' and 'right' modes.
Support exactly singular input.
Support rectangular input whether fat or thin.
QDWH and SVD polar decomp can be toggled.
Stability and performance improvements.
Jit everything.
Test tolerances adjusted.

Move tests to the test folder

Tests handle bfloat16 and multiple seeds

Fix linter errors, move polar into right namespace

Migrate eigh tests to tests folder

Move polar into _src.lax

Fix flake8 error

Fix test

Add license headers

Fix test failures on GPU/TPU

Another try at fixing accelerator tests

Fix tests on TPU?

Key dtypes on name

Disable half precision in tests

Skip eigh off CPU for now

adam-lewis-gcorp committed:
Adds spectral divide and conquer eigensolver

adam-lewis-gcorp commented:
@hawkinsp Added the eigh stuff - I can put some more tests in, but for now let's see what we can do about the Jitting

hawkinsp commented on jax/_src/scipy/test_eigh.py:
```diff
@@ -0,0 +1,49 @@
+"""Test for eigh.py."""
```
Would you mind moving the tests into the jax/tests directory?

adam-lewis-gcorp replied:
Done

hawkinsp commented on jax/_src/scipy/test_eigh.py:
```diff
@@ -0,0 +1,49 @@
+Ns = [16, 256]
+precisions = [lax.Precision.HIGHEST,]
+
+@pytest.mark.parametrize("N", Ns)
```
I'd avoid using pytest marks. We don't always use pytest to run tests (e.g., in our internal CI), so the tests need to work without pytest as the test runner, which means you can't use marks.

The test should work if you just run `python test_eigh.py` without pytest; you'd probably need to add a `__main__`  and move the tests into a class for this to work.

Typically we use absl's `parameterized` module for parametric tests.

Take a look at, say, `lax_numpy_test.py` for some examples.

adam-lewis-gcorp replied:
Done

hawkinsp commented on jax/_src/scipy/test_qdwh.py:
```diff
@@ -0,0 +1,95 @@
+  return result, spectrum
+
+
+@pytest.mark.parametrize("n_zero_svs", n_zero_svs)
```
Same here.

hawkinsp commented on jax/_src/scipy/eigh.py:
```diff
@@ -0,0 +1,223 @@
+  return evals, evecs
+
+
+def svd(A, precision=lax.Precision.HIGHEST):
```
It looks to me like the polar decomposition is jittable, correct?

One thing we could probably do is hook up the polar decomposition to the API (because why not, it's useful) and use the existing `eigh` implementation as the basis of an SVD on TPU, which is simply missing at the moment.

We can separately work on making the `eigh` above jittable.

adam-lewis-gcorp replied:
Yes, it should be jittable

hawkinsp commented on jax/_src/scipy/qdwh.py:
```diff
@@ -0,0 +1,235 @@
+"""
```
I'd be tempted to move this into `jax/_src/lax/qdwh.py`. This is mostly because we want to be able to call it from `jax.lax.linalg` primitives like SVD. Optional, though, I can do that in another PR.

adam-lewis-gcorp replied:
Moved

adam-lewis-gcorp committed:
Significant improvements to QDWH.

Mirror Scipy function signature.
Support both 'left' and 'right' modes.
Support exactly singular input.
Support rectangular input whether fat or thin.
QDWH and SVD polar decomp can be toggled.
Stability and performance improvements.
Jit everything.
Test tolerances adjusted.

adam-lewis-gcorp committed:
Move tests to the test folder

adam-lewis-gcorp committed:
Tests handle bfloat16 and multiple seeds

adam-lewis-gcorp committed:
Fix linter errors, move polar into right namespace

adam-lewis-gcorp committed:
Migrate eigh tests to tests folder

hawkinsp commented on tests/lax_scipy_test.py:
```diff
@@ -405,6 +454,137 @@ def testSphHarmCornerCaseWithWrongNmax(self):
+        side=side)
+      return
+
+    unitary, posdef, info = jsp.linalg.polar(matrix, method=method, side=side)
```
I'm actually missing where we export `polar` in the `jax.scipy.linalg` namespace. Don't we need to add it to the list in `jax/scipy/linalg.py`

adam-lewis-gcorp replied:
isn't it in there?

adam-lewis-gcorp committed:
Move polar into _src.lax

adam-lewis-gcorp committed:
Fix flake8 error

adam-lewis-gcorp commented on jax/scipy/linalg.py:
```diff
@@ -35,3 +35,8 @@
+from jax._src.lax.polar import (
+  polar,
+  polar_unitary
+)
```
The namespace adding is up here

hawkinsp replied:
Hmm I must have missed this.

adam-lewis-gcorp committed:
Fix test

hawkinsp reviewed (changes requested):
You need to add the Apache license header to the new files.

adam-lewis-gcorp committed:
Add license headers

adam-lewis-gcorp committed:
Fix test failures on GPU/TPU

adam-lewis-gcorp committed:
Another try at fixing accelerator tests

adam-lewis-gcorp committed:
Fix tests on TPU?

adam-lewis-gcorp committed:
Key dtypes on name

adam-lewis-gcorp committed:
Disable half precision in tests

adam-lewis-gcorp committed:
Skip eigh off CPU for now

adam-lewis-gcorp committed:
Merge branch 'master' of github.com:adam-lewis-gcorp/jax

hawkinsp commented:
Merged #7335 which is a squashed version of this PR.
